src/GraphCoarsen.cpp

```diff
@@ -13,788 +13,1100 @@
-    bool build_coarse_relation)
-{
-    // put trace_extensions and bubble_functions in Pedges
-    // the coarse dof numbering is as follows: first loop over each face, count
-    // the traces, then loop over each aggregate, count the bubble functions
-
-    const mfem::SparseMatrix& Agg_edge(graph_topology_.Agg_edge_);
-    const mfem::SparseMatrix& Agg_vertex(graph_topology_.Agg_vertex_);
-    const mfem::SparseMatrix& face_edge(graph_topology_.face_edge_);
-    const mfem::SparseMatrix& Agg_face(graph_topology_.Agg_face_);
-
-    const unsigned int nAggs = vertex_target.size();
-    const unsigned int nfaces = face_edge.Height();
-    const unsigned int nedges = Agg_edge.Width();
-
-    // element matrices for hybridization
-    if (build_coarse_relation)
-    {
-        CM_el.resize(nAggs);
+        SparseMatrix M_face = M_local.GetSubMatrix(edge_ext_dofs, edge_ext_dofs, col_marker_);
+        std::vector<double> M_diag = M_face.GetData();
+
+        sec_M.ReduceSend(face, std::move(M_diag));
     }
 
-    // construct face to coarse edge dof relation table
-    int* face_cdof_i = new int[nfaces + 1];
-    face_cdof_i[0] = 0;
-    for (unsigned int i = 0; i < nfaces; ++i)
-        face_cdof_i[i + 1] = face_cdof_i[i] + edge_traces[i].Width();
-    int total_num_traces = face_cdof_i[nfaces];
-    int* face_cdof_j = new int[total_num_traces];
-    std::iota(face_cdof_j, face_cdof_j + total_num_traces, 0);
-    double* face_cdof_data = new double[total_num_traces];
-    std::fill_n(face_cdof_data, total_num_traces, 1.0);
-    mfem::SparseMatrix newface_cdof(face_cdof_i, face_cdof_j, face_cdof_data,
-                                    nfaces, total_num_traces);
-    face_cdof.Swap(newface_cdof);
-
-    mfem::Vector local_rhs_trace, local_rhs_bubble, local_sol, trace, PV_trace;
-    mfem::Vector B_potential, F_potential;
-    int nlocal_fine_dofs, nlocal_coarse_dofs;
-    mfem::Array<int> local_fine_dofs;
-    mfem::Array<int> faces;
-
-    int* Agg_dof_i;
-    int* Agg_dof_j;
-    double* Agg_dof_d;
-    int Agg_dof_nnz = 0;
-    if (build_coarse_relation)
+    return sec_M.Collect();
+}
+
+void GraphCoarsen::ComputeEdgeTargets(const MixedMatrix& mgl,
+                                      const ParMatrix& face_perm_edge)
+{
+    const SparseMatrix& face_edge = face_perm_edge.GetDiag();
+
+    auto shared_sigma = CollectSigma(face_edge);
+    auto shared_M = CollectM(mgl.LocalM());
+    auto shared_D = CollectD(mgl.LocalD());
+
+    const SparseMatrix& face_shared = gt_.face_face_.GetOffd();
+
+    SharedEntityComm<DenseMatrix> sec_face(gt_.face_true_face_);
+    DenseMatrix collected_sigma;
+
+    int num_faces = gt_.face_edge_local_.Rows();
+
+    for (int face = 0; face < num_faces; ++face)
     {
-        Agg_dof_i = new int[nAggs + 1];
-        Agg_dof_i[0] = 0;
+        int num_face_edges = face_edge.RowSize(face);
+
+        if (!sec_face.IsOwnedByMe(face))
+        {
+            edge_targets_[face].SetSize(num_face_edges, 0);
```

Comment:
Is it right that this line is not needed?

Reply:
I think it was to make sure multiplying by a matrix would be well defined, even if the width is zero.  But this can be removed once `Mult` automatically resizes dense matrices.